### ISSUE TITLE
Segmentation fault on Dmend

### ISSUE THREAD

**Issue by JohnEarnest** [kevinlawler/kona#378]
In Kona:

    indigo:kona je$ ./k
    K Console - Enter \ for help
    
      .[3 3#0;(1 2;0 1);{x;y};1]
    Segmentation fault: 11

GDB indicates the crash is in `dot_tetradic_2()`, [here](https://github.com/kevinlawler/kona/blob/master/src/vd.c#L219)

For comparison, this works as expected in k2:

      .[3 3#0;(1 2;0 1);{x;y};1]
    (0 0 0
     1 1 0
     1 1 0)

**Comment by tavmem** [comment]
Progressively simpler cases in which Kona also fails (and then finally succeeds):
```
  .[3 3#0;(1 2;0 1);:;1]
(0 0 0
 0 0 0
 0 0 0)
valence error
.[3 3#0;(1 2;0 1);:;1]
                  ^
>  \
  
  
  
  .[3 3#0;(1;0 1);:;1]
(0 0 0
 0 0 0
 0 0 0)
valence error
.[3 3#0;(1;0 1);:;1]
                ^
>  \
  
  
  .[3 3#0;(1;0);:;1]
(0 0 0
 0 0 0
 0 0 0)
valence error
.[3 3#0;(1;0);:;1]
              ^
>  \
  
  
  .[3 3#0;(1);:;1]
(0 0 0
 0 0 0
 0 0 0)
valence error
.[3 3#0;(1);:;1]
            ^
>  \
  
  
  .[3 3#0;1;:;1]
(0 0 0
 1
 0 0 0)
```

**Comment by tavmem** [comment]
The monadic form segfaults in Kona for even the simplest case:
```
  .[3 3#0;(1 2;0 1);{x+1}]
Segmentation fault

  .[3 3#0;(1;0 1);{x+1}]
Segmentation fault

  .[3 3#0;(1;0);{x+1}]
Segmentation fault

  .[3 3#0;(1);{x+1}]
Segmentation fault

  .[3 3#0;1;{x+1}]
Segmentation fault
```

**Comment by tavmem** [comment]
There is also a different problem:
After the most recent merge for pull request 379 to
"Fix issues with the reverved parser not reading enough characters"
60 tests now fail.

Test pass rate: 0.9357, Total: 959, Passed: 873, Skipped: 26, Failed: 60, Time: 1.360000s
fail

Looks like that fix should be backed out.

**Comment by tavmem** [comment]
I just noticed the 2 outstanding pull requests.
1) Tests for 379 from @kirbyfan64
2) Fixes from @pahihu including a backout of 379

Since my backout of 379 now causes a conflict with @pahihu's backout, I will reverse mine and merge both requests.

**Comment by tavmem** [comment]
Not surprisingly, both of the @kirbyfan64 new tests fail after the @pahihu backout.
One fails in a trivial manner (reporting "parse" error, instead of "reserved" error.
Since k2.8 reports "parse" error for both tests, I will adjust the @kirbyfan64 test.

The other @kirbyfan64 test does not report an error at all.
That test needs some further research.
```
  (.[.:;,"{_vs_a[x]}";:]) 
(1
 "parse")
parse error
>  \
  
  (.[.:;,"{_vs.a[x]}";:])
(0;{_vs.a[x]})
```

**Comment by kirbyfan64** [comment]
@tavmem ACK! I'm sorry! I ran the tests, but I thought some were failing to begin with...I probably should've run `make clean`.

**Comment by tavmem** [comment]
Not a problem ... it happens.  Easily fixed by backing out the commit.
You should not have to run `make clean`.
The key is that ALL tests should pass.
They are regression tests ... no tests for missing features ... they should all pass by design.
And, whenever you add a feature, fix a bug, or resolve a crash, you should add a relevant test or tests.
Then future contributors can verify that their commits are not breaking anything.

**Comment by tavmem** [comment]
Back on issue topic:
Another simplified test that crashes:
```
  .[3 3#0;1;{x;y};1]
Segmentation fault

**Comment by tavmem** [comment]
Actually, the discussion of an adeguate set  of regression tests was very relevant to this issue:

Using the commit b332a1a07557d71409dd73f939a84abd92f21d6a of Apr 7 2015 you get
```
  .[3 3#0;(1 2;0 1);{x;y};1]
(0 0 0
 1 1 0
 1 1 0)
```
Using the commit 245f5cf71f5aa6ed2a67744ccf549330c3e479a7 of Apr 8 2015 you get
```
  .[3 3#0;(1 2;0 1);{x;y};1]
Segmentation fault
```
A better set of tests should have caught this regression.

**Comment by tavmem** [comment]
The commit 245f5cf of Apr 8 2015 provided more description on index error.
before:
```
  a:!6; a[9]:55
index error
> 
```
after:
```
  a:!6
0 1 2 3 4 5
  a[9]:55
index error
a[9]:55
    ^
>  \

  a:!6; a[9]:55
index error
a:!6; a[9]:55
at execution instance 2 of ":"
>
```
Somewhat ironically, the pull request #382 which merged 4 commits, eliminated all description on index error.  This is another regression.  Currently, Kona yields no message at all:
```
  a:!6; a[9]:55
  a
0 1 2 3 4 5
```
So, we have 2 levels of regression.

Next step: Try to figure out what the merged 4 commits in pull request #382 from @pahihu were attempting to accomplish.

**Comment by tavmem** [comment]
Commit f16c9aa34637d507b2cdb08d303962a74f64eded restores messages on index error.
This probably breaks the Derwish commit 6fe9e237e853953be0322186856d20d0dad72061 intended to reset fLoad/fCmplt on on fclose error.

Will need to follow up with a test case on the fclose error.
